Merge multiple IN predicates

pontusmelke opened the pull request:
For queries with multiple in on the same identifier separated by AND
we can merge into to one big list.

Examples:
```
MATCH (a) WHERE a.prop IN [1,2,3] AND a.prop IN [2,3,4] RETURN * =>
MATCH (a) WHERE a.prop IN [2,3]  RETURN *
```

```
MATCH (a) WHERE a.prop IN [1,2,3] AND a.prop IN [5,6,7] RETURN * =>
MATCH (a) WHERE FALSE RETURN *
```

```
MATCH (a) WHERE a.prop IN [1,2,3] OR a.prop IN [2,3,4] RETURN * =>
MATCH (a) WHERE a.prop IN [1,2,3,4]  RETURN *
```

pontusmelke committed:
Merge multiple IN predicates

For queries with multiple in on the same identifier separated by AND
we can merge to one big list.

fickludd commented on community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/mergeInPredicates.scala:
```diff
@@ -0,0 +1,82 @@
+//      In(Property(Variable(a),PropertyKeyName(prop)),ListLiteral(List(SignedDecimalIntegerLiteral(2), SignedDecimalIntegerLiteral(3)))),
+//      In(Property(Variable(a),PropertyKeyName(foo)),ListLiteral(List(StringLiteral(bar))))),
+//    In(Property(Variable(a),PropertyKeyName(prop)),ListLiteral(List(SignedDecimalIntegerLiteral(2), SignedDecimalIntegerLiteral(3))))),
+// In(Property(Variable(a),PropertyKeyName(foo)),ListLiteral(List(StringLiteral(bar)))))))), Return(false,ReturnItems(true,List()),None,None,None,None,Set()))))
```
Ghostlines!

fickludd commented on community/cypher/cypher-compiler-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/mergeInPredicatesTest.scala:
```diff
@@ -0,0 +1,76 @@
+  test("MATCH (a) WHERE (a.prop IN [1,2,3] AND a.prop IN [2,3,4]) OR (a.prop IN [2,3,4] AND a.prop IN [3,4,5]) RETURN *") {
+    shouldRewrite(
+      "MATCH (a) WHERE (a.prop IN [1,2,3] AND a.prop IN [2,3,4]) OR (a.prop IN [2,3,4] AND a.prop IN [3,4,5]) RETURN *",
+      "MATCH (a) WHERE a.prop IN [2,3] OR a.prop IN [3,4] RETURN *")
```
Could we rewrite 
`WHERE a.prop IN [2,3] OR a.prop IN [3,4]` => `WHERE a.prop IN [2,3,4]`?

fickludd reviewed (changes requested):
Some leftovers and a question. Like the concept. Thought the implementation was a little hard to read, but maybe I'm just inexperienced with rewriters, or they are just hard to code.

pontusmelke committed:
Add support for merging OR

For example
```
a IN [1,2] OR a IN [2,3]
```
will be rewritten to
```
a IN [1,2,3]
```

pontusmelke commented:
@fickludd 
- added support for `OR`, good idea!
- removed ghostly cruft
- Did very little to make the code more readable, not sure what to do actually? Open to suggestions.

pontusmelke committed:
No need to only rewrite only in where clauses.

fickludd commented:
Very nice @pontusmelke! Does this rewrite happen after `a.prop = 2` `=>` `a.prop IN [2]`? That would increase it's value.

pontusmelke commented:
@fickludd No unfortunately we do auto-parametrization in between, this rewriter depends having the values there for us to use. However I am investigating if we could move rewriters around a little bit since I think we do it wrong on a couple of places. Consider that a later optimization though.
